Implement size & layout checks for the glue layer

AndrejMitrovic opened the pull request:
As noted in the commit comment, I've tried to use friend declarations, but it's problematic due to circular #include's which end up breaking compilation in subtle ways (if only C++ had modules..).

Fixes #4

AndrejMitrovic committed:
Implement size & layout checks for the glue layer

This implements a partially-automatic way of
ensuring the size & layout of the structs on the
D and C++ size match.

Note: some types cannot be checked yet due to
the inability to access C++ private fields' offset
information (for example std::vector).

An attempt was made to use friend declarations
instead of making fields public on the C++ side.
This doesn't work as we end up with circular
includes which break compilation in various ways.

Fixes #4

Geod24 commented on source/scpd/tests/GlueTypes.d:
```diff
@@ -0,0 +1,74 @@
+    SCPQuorumSet,
+
+    /// scpd.types.Stellar_types
+    //Hash,  // todo (linker error)
```
`Hash` is aliased to `BitBlob!256`
But I would double check if they are binary compatible (esp. w.r.t. endianess)

AndrejMitrovic replied:
Ah I could just pass it as a void* then. Yeah I'll add those checks as well.

Geod24 replied:
Or you could pass it as a `ref Hash` on D side and `Hash&` on the C++ side?

Geod24 commented on source/scpd/tests/GlueTypes.d:
```diff
@@ -0,0 +1,74 @@
+    //Hash,  // todo (linker error)
+    //PublicKey,
+    //SignatureHint,  // todo: opaque_array!4 is not instantiated on the C++ side,
+                      // could be a type or mangling mismatch
```
https://github.com/stellar/stellar-core/search?q=SignatureHint&unscoped_q=SignatureHint

Geod24 commented:
Random idea: We could add `offsetoff` tests (or better... GENERATE THEM!)

Geod24 commented on source/scpp/extra/DSizeChecks.cpp:
```diff
@@ -0,0 +1,60 @@
+using namespace xdr;
+using namespace stellar;
+
+/// note: tried to use template but linking to it is a nightmare
```
Care to elaborate ?

AndrejMitrovic replied:
C++ side:

```C++
template <typename T>
unsigned long long cppSizeOf ( T &value ) { return sizeof(value); }

// instantiate
template unsigned long long cppSizeOf<Value>(Value &value);
```

D side:

```D
extern(C++) ulong cppSizeOf (T)(ref T);
alias cppSizeOf!Value cppSizeOf_;
```

Error:

```
Undefined symbols for architecture x86_64:
  "unsigned long cppSizeOf<xdr::xvector<unsigned char, 4294967292u> >(xdr::xvector<unsigned char, 4294967292u>&)", referenced from:
      __D4scpd5tests8SizeTest17__unittest_L31_C1FZv in agora-test-application.o
```

I'm not sure how well D links with C++ templates. Maybe I was missing something?

AndrejMitrovic commented:
> Random idea: We could add offsetoff tests (or better... GENERATE THEM!)

What do you mean? It already contains `.offsetof` tests.

Geod24 commented:
Right, I was looking at the `static assert` and thinking in addition to size it could also be `offsetof`. But that actually does not test anything... Disregard :D

AndrejMitrovic commented:
Btw, this is missing checks for passing structs around and testing functions. I will add those tests as well. First let me fix the failure on linux.

Geod24 commented:
Since you are testing offset and size, I'd say pass the struct by ref. Because if the size's wrong you might crash before reaching your test function.

AndrejMitrovic commented:
Pushed linux fix and added size & layout checks for the `Hash` type.

Edit: also rebased on v0.x.x

Geod24 commented on dub.json:
```diff
@@ -15,6 +15,8 @@
         "source/scpp/build/BallotProtocol.o",
         "source/scpp/build/ByteSliceHasher.o",
+        "source/scpp/build/DSizeChecks.o",
+        "source/scpp/build/DLayoutChecks.o",
```
Can we move this in tests only ?
See https://dub.pm/package-format-json.html#configurations

AndrejMitrovic replied:
Linking to the dub docs is cruel.

Geod24 commented on source/scpd/tests/GlueTypes.d:
```diff
@@ -0,0 +1,68 @@
+
```
Extra empty line

Geod24 replied:
Ping

Geod24 commented on source/scpd/tests/GlueTypes.d:
```diff
@@ -0,0 +1,68 @@
+/*******************************************************************************
+
+    Contains runtime size & ABI object layout checks and
+    function argument passing in the D glue layer works properly.
```
"that the D glue layer..." ?

Geod24 replied:
This just declares the types right ?

Geod24 commented on source/scpd/tests/LayoutTest.d:
```diff
@@ -0,0 +1,80 @@
+extern(C++) struct FieldInfo
+{
+    long size;
+    long offset;
```
So... Is `long` mangled correctly ? 🤔

AndrejMitrovic replied:
Yeah long seems fine. But not `unsigned lang lang`.

Geod24 commented on source/scpd/tests/LayoutTest.d:
```diff
@@ -0,0 +1,80 @@
+/// Special-case as we implement our own Hash type that is binary-compatible with SCP's Hash
+/// Note: cannot pass with proper type due to mangling issues:
+///  "isHashTypeCompatible(BitBlob<256ul>&)", referenced from:
+///      __D4scpd5tests10LayoutTest17__unittest_L43_C1FZv in agora-test-application.o
```
One question I have is: "how does mangling behave w.r.t `alias`. Aliases are transparent, so we should probably use `struct Name { RealType data; alias data this; }` however it has its problems (namely there's a Name -> RealType path but not the other way around).

Geod24 replied:
Yeah I think we might need to do this instead of aliases in the binding. Care to try ?

AndrejMitrovic replied:
Hmm, are there any benefits to this? Because implicit conversions will still happen. Genuinely curious as I rarely use subtyping.

Geod24 replied:
Well if you declare a `void fun(ref Hash)` function on the D side and a `void fun(Hash&)` on the C++ side and it doesn't link I'd say it's a pretty annoying bug.

AndrejMitrovic replied:
ok so you want me to try declaring Hash as:

```D
struct Hash
{
    BitBlob!(...) _internal;
    alias _internal this;
}
```

And see how far we can take this?

AndrejMitrovic committed:
Use a mangle-compatible definition of a Hash

This makes it possible to interface with C++ functions
which take or return a Hash type.

AndrejMitrovic commented:
Update: added a commit changing the `Hash` definition so it's mangle-compatible with C++. This means it can now be passed around to C++ functions.

I have yet to fix https://github.com/bpfkorea/agora/pull/9/files#r289269364 because dub documentation is terrible.

AndrejMitrovic commented:
Moved the layout & size C++ test support file sources to the `unittest` target, so they don't get unconditionally compiled & linked to.

Geod24 commented on source/scpd/tests/SizeTest.d:
```diff
@@ -0,0 +1,36 @@
+/*******************************************************************************
+
+    Contains runtime size & ABI object layout checks and
```
Just size tho

Geod24 commented on source/scpd/tests/LayoutTest.d:
```diff
@@ -0,0 +1,71 @@
+/*******************************************************************************
+
+    Contains runtime size & ABI object layout checks and
+    function argument passing in the D glue layer works properly.
```
Just layout tho

AndrejMitrovic commented:
Fixed docs and extra blank line (forgot to merge from another branch).

Geod24 commented on source/scpd/types/XDRBase.d:
```diff
@@ -88,34 +91,14 @@ class xdr_wrong_union : logic_error
-//struct xarray(T, uint32_t N) { public array!(T, N) data; alias data this; }
-struct xarray(T, uint32_t N)
+///
+struct opaque_array(uint32_t N = XDR_MAX_LEN)
```
It's only used for `Hash` right ?

AndrejMitrovic replied:
Also `uint256`, `SignatureHint` and some other types I guess. What are you thinking?

Geod24 reviewed (approved):
This is grrrrrreat

*auto-merge toggled on*
